src/access.rs

```diff
@@ -0,0 +1,361 @@
+use crate::error::LdkLiteError as Error;
+#[allow(unused_imports)]
+use crate::logger::{
+	log_error, log_given_level, log_info, log_internal, log_trace, log_warn, FilesystemLogger,
+	Logger,
+};
+
+use lightning::chain::chaininterface::{BroadcasterInterface, ConfirmationTarget, FeeEstimator};
+use lightning::chain::WatchedOutput;
+use lightning::chain::{Confirm, Filter};
+
+use bdk::blockchain::{Blockchain, EsploraBlockchain, GetBlockHash, GetHeight, GetTx};
+use bdk::database::BatchDatabase;
+use bdk::wallet::AddressIndex;
+use bdk::{SignOptions, SyncOptions};
+
+use bitcoin::{BlockHash, Script, Transaction, Txid};
+
+use std::collections::HashSet;
+use std::sync::{Arc, Mutex};
+
+/// The minimum feerate we are allowed to send, as specify by LDK.
+const MIN_FEERATE: u32 = 253;
+
+pub struct LdkLiteChainAccess<D>
+where
+	D: BatchDatabase,
+{
+	blockchain: EsploraBlockchain,
+	wallet: Mutex<bdk::Wallet<D>>,
+	queued_transactions: Mutex<Vec<Txid>>,
+	watched_transactions: Mutex<Vec<Txid>>,
+	queued_outputs: Mutex<Vec<WatchedOutput>>,
+	watched_outputs: Mutex<Vec<WatchedOutput>>,
+	last_sync_height: tokio::sync::Mutex<Option<u32>>,
+	logger: Arc<FilesystemLogger>,
+}
+
+impl<D> LdkLiteChainAccess<D>
+where
+	D: BatchDatabase,
+{
+	pub(crate) fn new(
+		blockchain: EsploraBlockchain, wallet: bdk::Wallet<D>, logger: Arc<FilesystemLogger>,
+	) -> Self {
+		let wallet = Mutex::new(wallet);
+		let watched_transactions = Mutex::new(Vec::new());
+		let queued_transactions = Mutex::new(Vec::new());
+		let watched_outputs = Mutex::new(Vec::new());
+		let queued_outputs = Mutex::new(Vec::new());
+		let last_sync_height = tokio::sync::Mutex::new(None);
+		Self {
+			blockchain,
+			wallet,
+			queued_transactions,
+			watched_transactions,
+			queued_outputs,
+			watched_outputs,
+			last_sync_height,
+			logger,
+		}
+	}
+
+	pub(crate) async fn sync_wallet(&self) -> Result<(), Error> {
+		let sync_options = SyncOptions { progress: None };
+
+		self.wallet
+			.lock()
+			.unwrap()
+			.sync(&self.blockchain, sync_options)
+			.map_err(|e| Error::Bdk(e))?;
+
+		Ok(())
+	}
+
+	pub(crate) async fn sync(&self, confirmables: Vec<&(dyn Confirm + Sync)>) -> Result<(), Error> {
```

Comment:
just for my understanding, top level function will eventually block on this right ? 
and that is before starting next sync call

Reply:
No, we spawn a new background task via the tokio runtime. This task runs in a loop however, currently set to sleep 5 secs between each sync (attempt).

Reply:
can this go wrong ? 
For e.g. if we start this sync every 1 sec and it takes 10 secs to complete each sync. Multiple syncs are active at the same instance of time.
Normally it would have been fine but since each sync acquires resource locks, multiple instances of this sync cannot run in parallel or independent of each other. 
In effect, they are running serially waiting for the previous one to complete first. This can create an ever growing waiting tasks queue if completionTime > syncTriggerInterval. 

Is this understanding correct ?

Reply:
No, no, it's just a single loop, so no concurrent sync attempts should be spawned, it would always wait 5 secs between finishing the last and starting the next attempt:

https://github.com/lightningdevkit/ldk-lite/blob/389f960d7533fcc3c57071d4dc15549d14de538b/src/lib.rs#L488-L524

Reply:
(The chosen parameters are however entirely up for debate)